File: TypeScript-DataStructures-and-Algorithm/src/ts/LinkedList.ts
import { defaultEquals, IEqualsFunction } from './util';
import {Node} from './models/linked-list-models'

export default class LinkedList<T> {

  protected count = 0;
  protected head: Node<T> | undefined;

  constructor(protected equalsFn: IEqualsFunction<T> = defaultEquals) {
  }

  push(element: T) {
    const node = new Node(element);
    let current;

    if (this.head == null) {
      this.head = node;
    } else {
      current = this.head;
      while (current.next != null) {
        current = current.next;
      }
      current.next = node;
    }
    this.count++;
  }

  getElementAt(index: number) {
    if (index >= 0 && index <= this.count) {
      let node = this.head;
      for (let i = 0; i < index && node != null; i++) {
        node = node.next;
      }
      return node;
    }
    return undefined;
  }

  insert(element: T, index: number): boolean {
    if (index >= 0 && index <= this.count) {
      const node = new Node(element);

      if (index === 0) {
        const current = this.head;
        node.next = current;
        this.head = node;
      } else {
        const prev = this.getElementAt(index - 1);
        node.next = prev?.next;
        // @ts-ignore
        prev?.next = node;
      }
      this.count++;
      return true;
    }
    return false;
  }

  removeAt(index: number) {
    if (index >= 0 && index <= this.count) {
      let current = this.head;

      if (index === 0) {
        const prev = this.getElementAt(index - 1);
        current = prev?.next;
        prev.next = current.next;
      }
      this.count--;
      return current?.element;
    }
    return undefined;
  }

  remove(element: T) {
    const index = this.indexOf(element);
    return this.removeAt(index);
  }

  indexOf(element: T) {
    let current = this.head;

    for (let i = 0; i <= this.count && current != null; i++) {
      if (this.equalsFn(element, current.element)) {
        return i;
      }
      current = current.next;
    }
    return -1;
  }

  isEmpty() {
    return this.size() === 0;
  }

  size() {
    return this.count;
  }

  getHead() {
    return this.head;
  }

  clear() {
    this.count = 0;
    this.head = undefined;
  }

  toString() {
    if (this.head == null) {
      return ' ';
    }
    let objString = `${this.head.element}`;
    let current = this.head.next;
    for (let i = 1; i < this.size() && current != null; i++) {
      objString = `${objString},${current.element}`;
      current = current.next;
    }
    return objString;
  }
}

File: Java-Data-Structures/src/main/java/LinkedList/LinkedList.java
package LinkedList;

public class LinkedList<E> {

    private class Node {
        private E element;
        private Node next;

        public Node(E element, Node next) {
            this.element = element;
            this.next = next;
        }

        public Node(E element) {
            this.element = element;
        }

        public Node() {
            this(null, null);
        }

        public String toString() {
            return element.toString();
        }

    }

    private Node dummyHead;
    private int size;

    public LinkedList() {
        dummyHead = new Node(null);
        size = 0;
    }

    public int getSize() {
        return size;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public void add(E element, int index) {
        if(index < 0 || index > size) {
            throw new IllegalArgumentException("Add failed. Illegal index.");
        }

        Node current = dummyHead;
        for (int i = 0; i < index; i ++) {
            current = current.next;
        }

        current.next = new Node(element, current.next);
        size ++;
    }

    public E get(int index) {
        if (index < 0 || index > size) {
            throw new IllegalArgumentException("Get failed. Illegal index.");
        }

        Node current = dummyHead;
        for (int i = 0; i < index; i ++) {
            current = current.next;
        }

        return current.element;
    }

    public void set(int index, E element) {
        if (index < 0 || index > size) {
            throw new IllegalArgumentException("Set failed. Illegal index.");
        }

        Node current = dummyHead;
        for (int i = 0; i < index; i ++) {
            current = current.next;
        }

        current.element = element;
    }

    private boolean contains(E element) {
        Node current = dummyHead.next;

        while (current != null) {
            if (current.element.equals(element))
                return true;
            current = current.next;
        }
        return false;
    }

    public E remove(int index) {
        if (index < 0 || index > size) {
            throw new IllegalArgumentException("Set failed. Illegal index.");
        }

        Node current = dummyHead;
        for (int i = 0; i < index; i ++) {
            current = current.next;
        }

        Node retNode = current.next;
        current.next = retNode.next;
        retNode.next = null;
        size --;

        return retNode.element;

    }

    public void removeElement(E element) {
        Node current = dummyHead;
        while (current.next != null) {
            if (current.element.equals(element))
                break;
            current = current.next;
        }
        if (current.next != null) {
            Node retNode = current.next;
            current.next = retNode.next;
            retNode.next = null;
            size --;
        }
    }

    @Override
    public String toString() {
        StringBuilder  stringBuilder = new StringBuilder();

        Node current = dummyHead.next;
        while (current != null) {
            stringBuilder.append(current + "->");
        }
        stringBuilder.append("NULL");

        return stringBuilder.toString();
    }

}
